### ISSUE TITLE
Minor optimisation of reducing array creation count.

### ISSUE THREAD

**Issue by kgrz** [Shopify/bootsnap#40]
I was going through the source code for the Cache module, and noticed [this line](https://github.com/Shopify/bootsnap/blob/master/lib/bootsnap/load_path_cache/cache.rb#L37):

```ruby
if [DOT_SO, *DL_EXTENSIONS].include?(ext)
```

This part is inside the `reduce` loop. Since this is an array of constants, I'm wondering if it would be better to initialise the array upfront, and then check for membership inside the loop, thereby avoiding the creation of the array caused by `*` operator on the `DL_EXTENSIONS` constant.

Do you think this micro optimisation is required? Or am I making a mistake somewhere and reading this wrong?

**Comment by burke** [comment]
I'm not against optimizing it, but it's in a context where:

1. An extra assignment would feel a little ugly (useless constant, or a local in a module context);
2. It's run only once on initial load, at a point where `$LOADED_FEATURES` is only about 100 items long, meaning this is only 100 allocations.

For these reasons, I wrote it the lazy way. I'd be surprised if it introduced more than a 5µs penalty.

Maybe a reasonable way to write this would be to:

```ruby
BUILTIN_FEATURES = begin
  native_extensions = [DOT_SO, *DL_EXTENSIONS]
  $LOADED_FEATURES.reduce({}) do |acc, feat|
    # ...
  end
end
```

The question is whether that's worth the extra complexity. I don't have a strong opinion.

**Comment by burke** [comment]
Going to close. I'd consider a PR but this is super low priority for me.
